### ISSUE TITLE
C preprocessor removes line breaks in comments

### ISSUE THREAD

**Issue by Gravelbones** [NationalSecurityAgency/ghidra#3779]
**Describe the bug**
When trying to parse C source and the include file contains // comments the linefeed on that line is removed from the output.
This causes lines to be very long and very hard to find errors in.

**To Reproduce**
1. Go to File
2. Click on Parse C Source
3. Setup input for a file with // as comment
4. Check CParserOutput.out for lines after a // in the file

**Expected behavior**
The // comment shouldn't remove linefeed

https://github.com/NationalSecurityAgency/ghidra/blob/436bb4873e8571b87a4d785bb828d9bf772b4867/Ghidra/Features/Base/src/main/javacc/ghidra/app/util/cparser/CPP/CPP.jj#L2213-L2214

For line 2213 it can be fixed by
```diff
- <_LCMT: "//" (~["\n","\r"])* ("\n"|"\r"|"\r\n")>
+ <_LCMT: "//" (~["\n","\r"])*>
```
But for longer comments (line 2214) another fix is needed: output any linebreaks found in the comment.

**Attachments**
I have included a zip with a simple .h file and the output produced.
[demozip.zip](https://github.com/NationalSecurityAgency/ghidra/files/7741211/demozip.zip)

**Environment (please complete the following information):**
 - OS: Windows 10
 - Java Version: 11.0
 - Ghidra Version: 10.0.4 PUBLIC
 - Ghidra Origin: official ghidra-sre.org distro,

**Comment by Pokechu22** [comment]
Your formatting got a bit messed up since markdown is parsing the `~` character for strikethrough; here's the original code (which I am formatting by posting a link to the relevant lines; GitHub automatically formats that):

https://github.com/NationalSecurityAgency/ghidra/blob/56ca81cfe85efec5a4f17a409f3f05fabf5f8e3f/Ghidra/Features/Base/src/main/javacc/ghidra/app/util/cparser/CPP/CPP.jj#L2219-L2223

And here's the new code, if I am understanding you correctly (which I am posting by using three backticks on either side):

```
SPECIAL_TOKEN :
{
  <_LINECOMMENT: \"//\" (~[\"\n\",\"\r\"])* >
  <_COMMENT: <CMT>(<ECMT>) (~["*"])* "*" ("*" | ~["*","/"] (~["*"])* "*")*  "/" >
}
```

And in diff form (by writing <code>\`\`\`diff</code> at the top and using <code>\`\`\`</code> on the bottom):

```diff
 SPECIAL_TOKEN :
 {
-  <_LINECOMMENT: \"//\" (~[\"\n\",\"\r\"])* (\"\n\"|\"\r\"|\"\r\n\")> |
+  <_LINECOMMENT: \"//\" (~[\"\n\",\"\r\"])* > |
   <_COMMENT: <CMT>(<ECMT>) (~["*"])* "*" ("*" | ~["*","/"] (~["*"])* "*")*  "/" >
 }
```

I note that there is also some similar code above, which might also need to be changed:

https://github.com/NationalSecurityAgency/ghidra/blob/56ca81cfe85efec5a4f17a409f3f05fabf5f8e3f/Ghidra/Features/Base/src/main/javacc/ghidra/app/util/cparser/CPP/CPP.jj#L2206-L2217

**Comment by Gravelbones** [comment]
Thanks for the hints and yes the problem is "bigger" than the change I have suggested.
Because multi line comments also eats line breaks it shouldn't eat, so line number reporting in the c parser is wrong.
For both type of comments any line breaks eaten up, should be output again, regardless of type, otherwise the c parser won't report line numbers correctly.
My fix will fix it for one line comments but not multi line comments.

**Comment by Gravelbones** [comment]
Did some more testing. The preprocessor is removing way to many linefeeds from the input file.
```
#if 0
Not here
#else 
Yes here
#endif
Mark1/*Comment with
linefeed */Mark2
//Line comment
END
```
Result
```
#line 1: ""
Yes here
Mark1Mark2
END
```
The error line number will go way of the mark, if the file is just a bit longer.

**Comment by emteere** [comment]
Just ran into this while fixing some other parsing issues for MSVC 2022 header files.
It pulled a "#pragma"  line onto the same line causing the C-parsing to fail.
I'll add this into the fixes.

We have been considering keeping the comments and not stripping them, so they could be parsed in the CParser and the comments could be added to the fields in a structure.  Maybe just keep the single end of line comments on each field as it could get dicey to put the comments in the right place since they are very free format.

**Comment by Gravelbones** [comment]
I have also been looking into this and found so many problems with the current code and the old unmaintained javacc compiler.
So I have basically started a complete rewrite of this to get something that is working on the old files I am trying to parse.
You can see some partial work here: https://github.com/Gravelbones/javacc21/tree/cpreproc
In the end, I think the 2 parsers currently being used, should be combined into one, which would give so many more options of doing things like you suggest.

**Comment by emteere** [comment]
This issue should be essentially addressed as described below in patch and master.

The PreProcessor will now try to keep as much information in the CparserPlugin.out file that is the result of the Pre-Processor parsing phase.

Lines that are #ifdef'ed out will have **'////'**-4 comment slashes in front of them.  Lines that are parsing directives that are evaluated will have **'///'**-3 comment slashes, and have their evaluation state at the end after **'==='**.
It isn't perfect, but hopefully with the added information, the exact line number won't matter as much anymore, although it is pretty close.

An attempt was made to keep more end of line comments for things like structures in hopes that the C Parsing phase could keep the comment on the fields of a structure.  However that proved difficult with so many strange things in various header files.  So it will have to be a future enhancement.

**Comment by ryanmkurtz** [comment]
Closed by 1f04e8c15fccbeb905169e0c60e0d465fa6636e1
